openmetadata-service/src/main/java/org/openmetadata/service/security/auth/BasicAuthenticator.java

```diff
@@ -0,0 +1,477 @@
+    if (token.getExpiryDate().compareTo(Instant.now().toEpochMilli()) < 0) {
+      throw new RuntimeException(
+          "Password Reset Token" + token.getToken() + "Expired token. Please issue a new request");
+    }
+    if (Boolean.FALSE.equals(token.getIsActive())) {
+      throw new RuntimeException("Password Reset Token" + token.getToken() + "Token was marked inactive");
+    }
+  }
+
+  public RefreshToken validateAndReturnNewRefresh(UUID currentUserId, TokenRefreshRequest tokenRefreshRequest)
+      throws JsonProcessingException {
+    String requestRefreshToken = tokenRefreshRequest.getRefreshToken();
+    RefreshToken storedRefreshToken = (RefreshToken) tokenRepository.findByToken(requestRefreshToken);
+    if (storedRefreshToken == null) {
+      throw new RuntimeException("Invalid Refresh Token");
+    }
+    if (storedRefreshToken.getExpiryDate().compareTo(Instant.now().toEpochMilli()) < 0) {
+      throw new RuntimeException("Expired token. Please login again : " + storedRefreshToken.getToken().toString());
+    }
+    // TODO: currently allow single login from a place, later multiple login can be added
+    // just delete the existing token
+    tokenRepository.deleteTokenByUserAndType(currentUserId.toString(), REFRESH_TOKEN.toString());
+    // we use rotating refresh token , generate new token
+    RefreshToken newRefreshToken = TokenUtil.getRefreshToken(currentUserId, UUID.randomUUID());
+    // save Refresh Token in Database
+    tokenRepository.insertToken(newRefreshToken);
+    return newRefreshToken;
+  }
+
+  private User getUserFromRegistrationRequest(RegistrationRequest create) {
+    String username = create.getEmail().split("@")[0];
+    String hashedPwd = BCrypt.withDefaults().hashToString(12, create.getPassword().toCharArray());
+
+    BasicAuthMechanism newAuthMechanism = new BasicAuthMechanism().withPassword(hashedPwd);
+    return new User()
+        .withId(UUID.randomUUID())
+        .withName(username)
+        .withFullyQualifiedName(username)
+        .withEmail(create.getEmail())
+        .withDisplayName(create.getFirstName() + create.getLastName())
+        .withIsBot(false)
+        .withUpdatedBy(username)
+        .withUpdatedAt(System.currentTimeMillis())
+        .withIsEmailVerified(false)
+        .withAuthenticationMechanism(
+            new AuthenticationMechanism()
+                .withAuthType(AuthenticationMechanism.AuthType.BASIC)
+                .withConfig(newAuthMechanism));
+  }
+
+  public void validateEmailAlreadyExists(String email) {
+    if (userRepository.checkEmailAlreadyExists(email)) {
+      throw new RuntimeException("User with Email Already Exists");
+    }
+  }
+
+  @Override
+  public JwtResponse loginUser(LoginRequest loginRequest) throws IOException, TemplateException {
+    String userName =
+        loginRequest.getEmail().contains("@") ? loginRequest.getEmail().split("@")[0] : loginRequest.getEmail();
```

Comment:
do we need to check this. JsonSchema should enforce the email must contain a email format

Reply:
we allow username login as well